### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by hisabimbola**
#### Summary

Implement the nodejs Boilerplate
#### Description

This PR is to create a Nodejs boilerplate according to our specifications. Any body that wants to start a Nodejs project in the company just needs to clone this repository and he/she is good to go.
It will also contain our guidelines to coding, process and best practices.
#### Reviewers

- [ ] Reviewer 1 has approved the PR
- [ ] Reviewer 2 has approved the PR
#### TODO
- Tests
  - [ ] Unit
  - [ ] Integration
  - [ ] Acceptance
- [ ] Documentation

### CODE REVIEW

**Comment by hisabimbola** [comment]
Just so that we don't have one big repo to review afterwards, opening this PR so that we review as development is on going.

**Comment by hisabimbola** [comment]
Also we have an NPM package for this on NPM [here](https://www.npmjs.com/package/commercetools-nodejs-boilerplate)

**Comment by Siilwyn** [comment]
Once we merge the NPM package should be '@commercetools/nodejs-boilerplate' right?

**Comment by hisabimbola** [comment]
@Siilwyn Not sure we have a final decision on that yet. That convention will affect all our repositories. So which ever we decide, we adjust accordingly. For now, it's just `commercetools-nodejs-boilerplate`
Also, I think it's _loosely related_ to your findings on lernaJS, whether we adopting it or not.

**Comment by Siilwyn** [comment]
Merged the 'initial-readme' branch so that when this PR is ready we can merge it and be done. :)

File: .istanbul.yml
Comment (PhilippSpo) [inline .istanbul.yml]: @hisabimbola do we still need this file?

Comment (hisabimbola) [reply .istanbul.yml]: Not anymore, I will remove it.

File: .travis.yml
Comment (emmenko) [inline .travis.yml]: I'm not sure this is really useful at the moment. As far as I know you need to provide the recipients. We usually use HipChat for build notifications. I would use that instead.

https://docs.travis-ci.com/user/notifications/#HipChat-notification

File: CONTRIBUTING.md
Comment (emmenko) [inline CONTRIBUTING.md:41]: 👍

File: README.md
Comment (emmenko) [inline README.md]: I think we can provide a bash script to prepare a "clean" installation of the boilerplate (e.g. that removes files specific to the boilerplate, install deps, etc).

Comment (emmenko) [inline README.md]: Shouldn't it be `npm run install` -> `npm install` ?

File: README.md
Comment (emmenko) [inline README.md]: Looks like this is pretty much the same as the "quick start" section.

File: README.md
Comment (emmenko) [inline README.md]: Why? I don't think this is a requirement.

File: package.json
Comment (emmenko) [inline package.json]: Why does it run coverage? I would expect that it just runs the tests and watches for changes. Coverage should stay a separate task.

File: package.json
Comment (emmenko) [inline package.json]: I think we should provide a version here

File: package.json
Comment (emmenko) [inline package.json]: Shouldn't matter much, but I would specify `commercetools` applications

File: package.json
Comment (hisabimbola) [reply package.json]: @emmenko coverage is a separate task, what this does is it runs reports coverage in html and that can be opened in the browser when developing for the dev to see.

File: package.json
Comment (emmenko) [inline package.json]: I'm not sure it makes much sense to publish this boilerplate to npm. If so, we should add `private: true`.

File: package.json
Comment (emmenko) [inline package.json]: I think we agreed that this should be `@commercetools/nodejs-boilerplate`.

File: package.json
Comment (hisabimbola) [reply package.json]: Published it for npm badges 😉  and for when we want to set up yeoman.

File: package.json
Comment (emmenko) [inline package.json]: We might consider having also a `package.json.template`, which contains placeholders for the library name and probably some common dependencies (babel, eslint, etc).

Or we keep the normal package.json but without much references to the boilerplate.

Comment (hisabimbola) [reply package.json]: Yeoman will really make more sense for this....

File: package.json
Comment (emmenko) [reply package.json]: But why is it under `test:watch` ?

We usually use this task in development, specifically when writing tests. I don't want to run the coverage at this stage. 

I'm also not sure about the `watch` task, if it's really necessary. We should keep one or the other.
And usually coverage is something a build service runs.

What do you guys think?

File: .travis.yml
Comment (Siilwyn) [reply .travis.yml]: It works, but I was surprised it does. Really don't like getting emails about it though so Hipchat notifications sound good. I can't really figure out how to make it work though, @hisabimbola could you take a look at this?

File: README.md
Comment (Siilwyn) [reply README.md]: True, I merged the two readme's we had and wanted to keep both for reference. Just put them together. ;)

File: package.json
Comment (Siilwyn) [reply package.json]: Running only the tests while developing sounds good to me, for the coverage options I would prefer just the CLI output.

File: .travis.yml
Comment (emmenko) [reply .travis.yml]: You need to get an hipchat API token first, and the hipchat room ID. Then you encrypt those with travis CLI and you add the encrypted token in this config.

**Review by emmenko — CHANGES_REQUESTED** [review]
> Testing the new github feature ;)

Please address the comments first before merging.

**Comment by emmenko** [comment]
Uuh, you can see the list of conflicting files now. Nice ♥️

File: package.json
Comment (PhilippSpo) [reply package.json]: I agree with @Siilwyn and @emmenko here

File: .travis.yml
Comment (hisabimbola) [reply .travis.yml]: I do not have enough permission to add hipchat notification to a room. I will talk to Hajo to do that.

File: README.template.md
Comment (Siilwyn) [inline README.template.md]: I would say leave both the contributing and license part out, I think it only adds obvious information. Especially for the license since GitHub now also shows it in the top bar of the project.

File: README.md
Comment (Siilwyn) [reply README.md]: I'm creating a bash script that clean everything.

File: package.json
Comment (emmenko) [inline package.json]: Let's just be consistent and keep only "First + last name, email"

File: package.json
Comment (emmenko) [inline package.json]: Same here

Comment (emmenko) [reply package.json]: Also use your commercetools email

File: package.json
Comment (emmenko) [inline package.json]: This doesn't exist anymore I think

File: package.template.json
Comment (emmenko) [inline package.template.json:2]: Nitpick: keep the placeholder camelcase, it's easier to select/replace it ;)

Comment (emmenko) [reply package.template.json:2]: Ooh I see you do it in the setup script. Nice 👍

File: README.md
Comment (PhilippSpo) [inline README.md:17]: What about tests?

File: README.md
Comment (PhilippSpo) [inline README.md]: `instalÂl`? 😜

File: README.template.md
Comment (PhilippSpo) [reply README.template.md]: I agree with the License. I would be fine with keeping the contributing document here, though.

File: test/.eslintrc
Comment (PhilippSpo) [inline test/.eslintrc]: @hisabimbola we can fix this now. The glob feature was released in version `1.15.0` https://github.com/benmosher/eslint-plugin-import/blob/master/CHANGELOG.md.

File: package.template.json
Comment (PhilippSpo) [reply package.template.json:2]: yes very nice setup script. However we still need to mention it in the readme right?

File: .babelrc
Comment (PhilippSpo) [inline .babelrc:5]: @hisabimbola Is this still needed?

Comment (hisabimbola) [reply .babelrc:5]: Yes, [NYC](https://github.com/istanbuljs/nyc#use-with-babel-plugin-istanbul-for-es6es7es2015-support) uses it

Comment (PhilippSpo) [reply .babelrc:5]: Ok thanks for clarifying 👍

File: README.template.md
Comment (hisabimbola) [reply README.template.md]: Fixed

File: package.json
Comment (hisabimbola) [reply package.json]: Done.

File: package.json
Comment (hisabimbola) [reply package.json]: Updated.

File: README.md
Comment (hisabimbola) [reply README.md:17]: Added.

File: README.md
Comment (hisabimbola) [reply README.md]: Done.

File: test/.eslintrc
Comment (hisabimbola) [reply test/.eslintrc]: Fixed in the [eslint-config module](https://github.com/commercetools/eslint-config/pull/9)

**Comment by JoshBones** [comment]
I don't know if we care, but since we're using `cross-env` I thought I'd mention that some of the scripts won't work as is in windows

**Review by PhilippSpo — APPROVED** [review]

**Comment by hisabimbola** [comment]
@JoshBones 💭 interesting to know, like which scripts won't work?? [cross-env](https://github.com/kentcdodds/cross-env#cross-env) boasts that it works in windows

File: README.md
Comment (emmenko) [inline README.md]: Should we specify which min version?

File: CONTRIBUTING.md
Comment (emmenko) [inline CONTRIBUTING.md]: This should just be `npm start`

**Comment by JoshBones** [comment]
@hisabimbola `cross-env` seems to work great for setting the environment vars, but something like the lint script `./scripts/lint` doesn't seem to work since windows doesn't run scripts like that.

```
'.' is not recognized as an internal or external command, operable program or batch file.
```

That aside, even if it did bash scripts just won't work on windows anyway (at least without installing some kind of bash shell)

we could get around it by using js instead of sh, but that could be more trouble than its worth considering we don't even use windows 😛 

**edit** actually [this](http://eslint.org/docs/developer-guide/nodejs-api#cliengine) could make the lint script easy enough to port to js

File: README.md
Comment (Siilwyn) [reply README.md]: Add a non-windows machine as a requirement. 📦

File: README.md
Comment (emmenko) [inline README.md]: We should mention here the `setup` script and reformulate the steps a bit.

**Comment by emmenko** [comment]
@JoshBones thanks for the feedback, but we don't care much for windows at the moment. I wouldn't consider it much of a problem anyway since this is for developing a library.

I would worry about that only for runtime eventually, and if someone asks to make it run on windows. Until then I wouldn't worry much ;)

File: README.md
Comment (emmenko) [inline README.md]: Does is make sense to mention this if we're not using it here and we don't know if we're going to use it later?

File: README.md
Comment (emmenko) [inline README.md]: What is this section for?

File: test/main.spec.js
Comment (emmenko) [inline test/main.spec.js]: To simplify a bit this, we can add `src` to the `NODE_PATH`.

See https://github.com/sphereio/sphere-node-sdk/blob/rewrite-2.0/package.json#L39

Comment (Siilwyn) [reply test/main.spec.js]: How would you use a variable in 'require'?

Comment (emmenko) [reply test/main.spec.js]: The import then will be `import addTwoNumbers from 'main'`, the path to `src` is then implicit in the `NODE_PATH`

Comment (hisabimbola) [reply test/main.spec.js]: @emmenko [eslint-plugin-import](https://github.com/benmosher/eslint-plugin-import) does not support [NODE_PATH natively](https://github.com/benmosher/eslint-plugin-import/issues/130), to make it to work, we need to add some [new settings](https://github.com/benmosher/eslint-plugin-import/issues/130#issuecomment-160416915) in a .eslintrc file. Should we do that at the [eslint-config](https://github.com/commercetools/eslint-config) plugin level or just add the file to the nodejs-boilerplate?
Cc @Siilwyn @PhilippSpo

File: README.md
Comment (hisabimbola) [reply README.md]: @Siilwyn ?

File: README.md
Comment (hisabimbola) [reply README.md]: @Siilwyn is working on something here

File: README.md
Comment (hisabimbola) [reply README.md]: @Siilwyn ...

File: CONTRIBUTING.md
Comment (hisabimbola) [reply CONTRIBUTING.md]: `npm start` sounds like a server is starting up or something....

Comment (hisabimbola) [reply CONTRIBUTING.md]: Done.

File: test/main.spec.js
Comment (PhilippSpo) [reply test/main.spec.js]: I vote for adding a new setting to the local eslint config (not the commercetools-eslint-config), because this feels very application specific.

File: README.md
Comment (PhilippSpo) [inline README.md]: what is the optional second directory for?

Comment (Siilwyn) [reply README.md]: Should I clarify it more. For example if the project has a CLI you would pass in bin like:
`sh setup.sh awesome-project bin`

Comment (PhilippSpo) [reply README.md]: And what will it do with this information? If I want a bin folder I create one later, right?

Comment (Siilwyn) [reply README.md]: It will create the directory and add it to the watch task. If you omit the argument the setup script will still work.

File: test/main.spec.js
Comment (emmenko) [reply test/main.spec.js]: Yes (sorry I forgot about that), just in the package.json of the application itself, not in the eslint config.

Comment (hisabimbola) [reply test/main.spec.js]: Done.

File: package.json
Comment (ebonertz) [inline package.json:49]: How is this related to the node path variable? Is this related to Babel?

File: README.md
Comment (ebonertz) [inline README.md]: "You can pass the project name and an additional directory name that you want to use, for example 'bin' "

**Review by ebonertz — APPROVED** [review]
Looks great! Almost ready for testing and use.

File: README.md
Comment (Siilwyn) [reply README.md]: That's a lot more descriptive, roll with this one @hisabimbola ?

Comment (hisabimbola) [reply README.md]: Yeah @Siilwyn, look good.

Long time goal though is for us to use Yeoman, will really reduce the need for templates...

File: README.md
Comment (PhilippSpo) [reply README.md]: Ah ok, so the adding to the watch task is what it really gives you. Maybe we can explain this a bit better, by saying that the additional directory will be automatically added to the watch task.

File: package.json
Comment (PhilippSpo) [reply package.json:49]: This is not related to Babel. In the tests, we set the `NODE_PATH` variable which lets us import from `src` instead of having to reference it all the time (like `../../../src/...`).
There is, however, an eslint plugin called `eslint-plugin-import` which also resolves your import statements and checks whether they are valid. Thus we also need to tell this plugin, that we can also import/resolve from `src` directly.

File: .travis.yml
Comment (PhilippSpo) [inline .travis.yml:14]: 👍

**Review by emmenko — APPROVED** [review]

**Comment by emmenko** [comment]
Ready to go from my side. Good work guys! 👍 🎉

**Comment by Siilwyn** [comment]
⚡ :shipit:

**Comment by PhilippSpo** [comment]
🎉🚀

**Pull request merged**
